src/transform/src/symbolic.rs

```diff
@@ -0,0 +1,612 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Definition and trait instances for working with symbolic algebraic expressions, [`SymbolicExpression`]
+
+use mz_ore::cast::CastLossy;
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use ordered_float::OrderedFloat;
+
+/// Symbolic algebraic expressions over variables `V`
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum SymbolicExpression<V> {
+    /// A constant expression
+    Constant(OrderedFloat<f64>),
+    /// `Variable(x, n)` represents `x^n`
+    Symbolic(V, usize),
```

Comment:
`-1` would be division, but `usize` excludes this. There seem to be also no other way here to express division, right? This came to mind because Section 2.3 in [this paper](https://www.vldb.org/pvldb/vol9/p204-leis.pdf) has a formula with a division.

Reply:
Right now I interpret `FactorizerVariable::Index` as the inverse, i.e., 1/|# of distinct key values|, which seems to be enough for our purposees.

There's definitely a lot of room for improvement in the symbolic algebra: we're not doing any of the simplification algorithms. I'm happy to implement them, but I'm not sure it's a great use of time.